File: py/control.py
#!/usr/bin/env python

import argparse
import common
import logging
import math
import sys

from collections import defaultdict
from threading import Thread
from time import time
from urllib.request import urlopen


class ControlHandler(common.BaseHandler):
    def do_GET(self):
        """ Get the current value of the counter.

        Poll all our backend servers and see if we have majority consensus.
        Sends a 200 and the value to the client if we have a consensus, 500 otherwise.
        """
        if not self.validate_or_reject_path():
            return
        value = self.server.get_value_from_vaults()
        status_code = 200 if value >= 0 else 500
        body = f'{value}'.encode('utf-8')
        self.send_status_and_response(status_code, 'text/plain', body)

    def do_POST(self):
        """ Update the value in storage to what is provided in the body.
        """
        if not self.validate_or_reject_path():
            return
        post_body_bytes = self.get_post_body_bytes()
        if not post_body_bytes:
            self.send_status_and_response(400, 'text/plain', 'Invalid or missing POST body')
            return
        [status_code, update_bytes] = self.server.send_value_to_vaults(post_body_bytes)
        self.send_status_and_response(status_code, 'text/plain', update_bytes)


class VaultUpdater(Thread):
    """ Vault thread to send an updated value to a single storage vault.

    We subclass from the Thread object so we can pass back the response (if any) we get from the vault.
    """
    def __init__(self, vault, value, timeout_sec=1):
        Thread.__init__(self)
        self.vault = vault
        self.value = value.encode('utf-8') if isinstance(value, str) else value
        self.timeout_sec = timeout_sec
        self._response = None

    def run(self):
        logging.debug(f'Setting vault {self.vault} value {self.value}')
        try:
            with urlopen(f'http://{self.vault}/', self.value, timeout=self.timeout_sec) as f:
                response = f.read()
                if response is not None and response:
                    self._response = int(response.decode('utf-8'))
        except Exception as error:
            logging.warning(f'Error setting vault {self.vault} value to {self.value}: {error}')

    def response(self):
        return self._response


class ControlServer(common.BaseServer):
    """ Control server which manages all our vaults and is the single point of contact for all reads and writes.
    """
    def __init__(self, address='', port=8000, vaults=None):
        super(ControlServer, self).__init__(address=address, port=port, handler_class=ControlHandler)
        self._vaults = []
        self._min_value = 0
        self._set_vaults(vaults)
        logging.info(f'Defined {len(self._vaults)} vaults')

    def vaults(self):
        """ Get the list of vaults on which we're storing information.
        """
        return self._vaults

    def send_value_to_vaults(self, value):
        """ Store a value on each of the vaults specified in the constructor.

        We spawn an updater thread for each vault so we won't get slowed down by an unreachable or slow vault.
        Let the client know how many vaults were successfully updated.
        """
        n = int(value.decode('utf-8'))
        if n < self._min_value:
            msg = f'Client would make value decrease from {self._min_value} to {n}'
            logging.warning(msg)
            return [400, msg]
        count = 0
        threads = []
        for vault in self._vaults:
            thread = VaultUpdater(vault, value)
            thread.start()
            threads.append(thread)
        for thread in threads:
            thread.join()
            count = count + 1 if thread.response() is not None else count
        status_code = 200
        if not self._has_majority(count):
            # If we don't have a majority, this is an error on the storage infrastructure. Indicate that we
            # do not have confidence that the storage request was successful.
            status_code = 500
        else:
            self._min_value = n
        return [status_code, f'Sent updates to {count}/{len(self._vaults)} vaults'.encode('utf-8')]

    def get_value_from_vaults(self):
        """ Get the consensus value stored across our vaults.

        Talk to each vault and get the value stored in said vault. If a majority of the vaults have the same
        value, then we have consensus and can return that value. If there is no consensus, return -1.
        """
        # The 'counts' dictionary stores a map from a possible value to the number of vaults which have that value.
        counts = defaultdict(int)
        for vault in self._vaults:
            value = self._get_value_from_vault(vault)
            if value is not None:
                logging.debug(f'Get vault {vault} Value {value}')
                counts[value] += 1
        logging.info(f'Counts data: {counts}')
        if not counts:
            # We got back no data at all from the vaults. Maybe they're all unreachable?
            logging.error('Could not reach any vaults to get counts data')
            return -1
        # How many vaults store the most common value(s)?
        # Note that there could be more than one value which is the "most common" value;
        # E.g., if we have seven vaults, and:
        # - vaults (A, C, G) have value "1";
        # - vaults (B, D, E) have value "2"; and
        # - vault F has value "4"
        # then the maximum number of vaults with the same value is three (the first two groups).
        max_vaults_with_common_value = max(counts.values())
        if not self._has_majority(max_vaults_with_common_value):
            # The largest grouping of vaults with a common value is not enough to reach consensus.
            logging.warning(f'No majority; only have {max_vaults_with_common_value}/{len(self._vaults)} with a consensus value')
            return -1
        # Return the value which represents consensus.
        return max([v for v,c in counts.items() if c == max_vaults_with_common_value])

    def _set_vaults(self, vaults):
        """ Set the addresses of each vault we're going to use to store information.
        """
        if vaults is None or not vaults:
            return
        self._vaults = vaults.strip().split(',') if vaults.strip() else []
        self._vaults = [w for w in self._vaults if w]

    def _has_majority(self, max_vaults_with_common_value):
        """ Is the largest grouping of vaults with a common value enough to represent a >50% majority?
        """
        num_vaults = len(self._vaults)
        num_for_majority = math.floor(num_vaults / 2) + 1
        return max_vaults_with_common_value >= num_for_majority

    def _get_value_from_vault(self, vault, timeout_sec=1):
        """ Fetch the current value from a given vault, with a timeout.
        """
        try:
            with urlopen(f'http://{vault}/', timeout=timeout_sec) as f:
                v = f.read()
                if v is not None and v:
                    return int(v.decode('utf-8'))
            return None
        except Exception as error:
            logging.warning(f'Error getting value from vault {vault}: {error}')
        return None


def main(argv):
    parser = common.make_parser()
    parser.add_argument('--vaults', type=str, default='', help='Comma-separated list of vaults')
    parsed = parser.parse_args()
    logging.getLogger().setLevel(parsed.loglevel)
    with ControlServer(port=parsed.port, vaults=parsed.vaults) as control:
        control.serve_forever()
    return 0


if __name__ == '__main__':
    sys.exit(main(sys.argv))

File: go/control.go
package main

import (
	"bytes"
	"errors"
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/antithesishq/antithesis-sdk-go/assert"
	"github.com/antithesishq/antithesis-sdk-go/lifecycle"
	"github.com/golang/glog"
)

type Details map[string]any

// A control server which maintains a list of vaults which will store the data.
type ControlServer struct {
	mux      *http.ServeMux
	Vaults   []string
	minValue int
	lock     sync.RWMutex
}

//go:generate antithesis-go-generator -v antithesis.com/go/glitch-grid

// Create and return a new Control server instance.
// Provide a comma-separated list of vaults with which we will communicate.
func NewControlServer(vaults string) *ControlServer {
	assert.Always("Instantiates a Control Server", true, nil)
	s := new(ControlServer)
	s.mux = http.NewServeMux()
	s.Vaults = strings.Split(vaults, ",")
	s.minValue = 0
	s.lock = sync.RWMutex{}
	s.mux.HandleFunc("/", s.handle)
	// Set the default timeout for all HTTP operations to be one second.
	http.DefaultClient.Timeout = time.Second
	glog.Infof("Defined %d vaults", len(s.Vaults))
	if len(s.Vaults) == 23456789 {
		assert.Unreachable("We have 23456789 vaults should be unreachable", Details{"numVaults": len(s.Vaults)})
		
		assert.Always("This line should never execute, but since this is an always assert, it will fail in Antithesis.", true, nil)
		assert.Reachable("This line should never execute, but since this is a reachable assert, it will fail in Antithesis.", Details{"numVaults": len(s.Vaults)})
	}
	assert.Reachable("Always returns a ControlServer when requested", Details{"vaults": vaults, "numVaults": len(s.Vaults)})
	return s
}

// Handle GET and POST requests to the root path.
func (s *ControlServer) handle(w http.ResponseWriter, r *http.Request) {
	if r.URL.Path != "/" {
		assert.AlwaysOrUnreachable("Control service: received a non-root request paths & handled that correctly.", true, Details{"path": r.URL.Path})
		// We only support operations on the root path.
		http.NotFound(w, r)
		return
	}
	if r.Method == http.MethodGet {
		s.get(w, r)
	} else if r.Method == http.MethodPost {
		s.post(w, r)
	} else {
		assert.AlwaysOrUnreachable("Control service: received a http method that is not a GET or a POST & handled that correctly.", true, Details{"method": r.Method})
		// Do not support PATCH, DELETE, etc, operations.
		http.NotFound(w, r)
	}
}

// Get the current value of the counter.
// Poll all our backend servers and see if we have majority consensus.
// Sends a 200 and the value to the client if we have a consensus, 500 otherwise.
func (s *ControlServer) get(w http.ResponseWriter, r *http.Request) {
	assert.Always("Control service: received a request to retrieve the counter's value", true, nil)
	result := s.getValueFromVaults()
	var statusCode int
	var body string
	if result >= 0 {
		assert.AlwaysOrUnreachable("Counter's value retrieved", true, Details{"counter": body, "status": statusCode})
		statusCode = http.StatusOK
		body = fmt.Sprintf("%d", result)
	} else {
		assert.Unreachable("Counter should never be unavailable", Details{"result": result})
		statusCode = http.StatusInternalServerError
		body = "-1"
	}

	expected_status := (statusCode == http.StatusOK) || (statusCode == http.StatusInternalServerError)
	assert.AlwaysOrUnreachable("HTTP return status is expected", expected_status, Details{"status": statusCode})
	assert.Always("The server never return a 500 HTTP response code", statusCode != http.StatusInternalServerError, Details{"status": statusCode})
	w.WriteHeader(statusCode)
	w.Write([]byte(body))
}

// Get the consensus value stored across our vaults.
// Talk to each vault and get the value stored in said vault. If a majority of the vaults have the same
// value, then we have consensus and can return that value. If there is no consensus, return -1.
func (s *ControlServer) getValueFromVaults() int {
	var wg sync.WaitGroup
	m := sync.RWMutex{}
	// Map from a value to the number of vaults which currently have that value.
	counts := map[int]int{}
	// Loop over all the vault addresses, and execute each one in a separate goroutine.
	// Use a WaitGroup to keep track of the pending functions, and a ReadWrite lock to
	// protect access to the counts tracker.
	for _, vault := range s.Vaults {
		wg.Add(1)
		go func(m *sync.RWMutex, vault string, counts map[int]int) {
			defer wg.Done()
			getValueFromVault(m, vault, counts)
		}(&m, vault, counts)
	}
	wg.Wait()
	glog.Infof("Counts data: %v", counts)
	if len(counts) == 0 {
		glog.Error("Could not reach any vaults to get counts data")
		return -1
	}
	// Iterate over the map of values to the count of vaults with that value.
	// If any count represents a majority, then by default it will have the maximum
	// number of vaults associated with it. Otherwise, just keep track the maximum
	// number of counts associated with any value.
	// E.g., if we have seven vaults, and:
	// - vaults (A, C, G) have value "1";
	// - vaults (B, D, E) have value "2"; and
	// - vault F has value "4"
	// then the maximum number of vaults with the same value is three (the first two groups),
	// but is not enough to achieve consensus.
	maxVal := 0
	for v, c := range counts {
		if c > maxVal {
			maxVal = c
		}
		if s.hasMajority(c) {
			// We have consensus. Return the value.
			return v
		}
	}
	// We do not have consensus, but we do know how popular the most common value(s) is/are.
	glog.Warningf("No majority; only have %d/%d with a consensus value", maxVal, len(s.Vaults))
	return -1
}

// Get the value stored in a single vault.
// If we are able to fetch a valid integer from the vault, update the counts map with that
// information in a thread-safe way. Otherwise, return without updating (but log the issue).
func getValueFromVault(m *sync.RWMutex, vault string, counts map[int]int) {
	url := fmt.Sprintf("http://%s/", vault)
	var resp *http.Response
	var err error
	if resp, err = http.Get(url); err != nil {
		// This could include a timeout.
		glog.Warningf("Error getting value from vault %s: %v\n", url, err)
		return
	}
	if resp.StatusCode != http.StatusOK {
		// Vault was not happy.
		glog.Warningf("Error getting value from vault %s: invalid status code %v\n", url, resp.StatusCode)
		return
	}
	body, readError := ioutil.ReadAll(resp.Body)
	if readError != nil {
		// Vault was supposedly-happy but did not return a value.
		glog.Warningf("Error getting value from vault %s: error reading from body: %v\n", url, readError)
		return
	}
	v, e := strconv.Atoi(string(body))
	if e != nil {
		// Vault returned a value, but it was not a valid integer.
		glog.Warningf("Error getting value from vault %s: invalid body response: %v (%v)\n", url, body, e)
		return
	}
	// If we've gotten here, then we received a valid integer back from the vault.
	// Start the map manipulation operation critical section.
	m.Lock()
	count, ok := counts[v]
	if !ok {
		// This value is not (yet) in the map. IOW, there are currently 0 vaults storing that value.
		count = 0
	}
	counts[v] = count + 1
	m.Unlock()
	// End of the map manipulation critical section.
	glog.V(1).Infof("Get vault %s Value %d", url, v)
}

// TODO: Call this when we detect that a vault is in a bad state.
func healFailingVault(vault string) {
	assert.Sometimes("Control service: invoked heal function on unhealthy vault", true, Details{"vault": vault})

	// Code to heal a failing vault
}

// Update the value in storage to what is provided in the body.
// Contact each vault and store that value in the vault.
func (s *ControlServer) post(w http.ResponseWriter, r *http.Request) {
	body, err := io.ReadAll(r.Body)
	if err != nil {
		// We did not get a valid body from the client. Tell them so.
		glog.Warningf("Could not read body: %v\n", err)
		w.WriteHeader(http.StatusBadRequest)
		w.Write([]byte("Invalid or missing POST body"))
		return
	}
	n, e := strconv.Atoi(string(body))
	if n < 0 || e != nil {
		// We got a body, but it is not a valid integer (or not valid for us).
		w.WriteHeader(http.StatusBadRequest)
		w.Write([]byte("Invalid or missing POST body"))
		return
	}
	// Check to make sure that this value is larger than the one we've previously committed
	s.lock.RLock()
	if n < s.minValue {
		msg := fmt.Sprintf("Client would make value decrease from %d to %d", s.minValue, n)
		s.lock.RUnlock()
		glog.Warning(msg)
		w.WriteHeader(http.StatusBadRequest)
		w.Write([]byte(msg))
		return
	}
	s.lock.RUnlock()
	// Send the update to the vaults, keeping track of how many vaults actually responded to us.
	// Technically this is a set(), but because Go doesn't have sets, this is a map of vaults to
	// booleans, where the value stored in the map doesn't really matter. The presence of ANY
	// value is enough to show that we got a successful response from the vault.
	resp := make(map[string]bool)
	assert.AlwaysOrUnreachable("Control service: there are vaults to update", len(s.Vaults) > 0, Details{"numVaults": len(s.Vaults)})
	s.postValueToVaults(body, resp)
	// If the number of responses represents a majority of the vaults, then we can claim success
	// in storing this value in our system. Otherwise it represents a server failure.
	if s.hasMajority(len(resp)) {
		w.WriteHeader(http.StatusOK)
		// Set the min value here to prevent us from going backwards.
		s.lock.Lock()
		assert.AlwaysOrUnreachable("Control service: unnecessary update attempted",
			n > s.minValue,
			Details{"minValue": s.minValue, "requestedValue": n})
		s.minValue = n
		s.lock.Unlock()
	} else {
		w.WriteHeader(http.StatusInternalServerError)
	}
	// In addition to the status code, unconditionally return a message of how many vaults we updated.
	w.Write([]byte(fmt.Sprintf("Sent updates to %d/%d vaults", len(resp), len(s.Vaults))))
}

// Actually send the POST commands to the vaults.
func (s *ControlServer) postValueToVaults(body []byte, resp map[string]bool) {
	// Use a WaitGroup so we can run the requests in parallel goroutine threads.
	var wg sync.WaitGroup
	// We will need to synchronize access to the response map.
	m := sync.RWMutex{}
	// For each vault, send a POST message containing the same body we received from the client.
	for _, vault := range s.Vaults {
		wg.Add(1)
		go func(m *sync.RWMutex, vault string, body []byte, resp map[string]bool) {
			defer wg.Done()
			glog.V(1).Infof("Setting vault %s value to %s", vault, string(body))
			url := fmt.Sprintf("http://%s/", vault)
			r, err := http.Post(url, "text/plain", bytes.NewBuffer(body))
			if err == nil && r != nil && r.StatusCode == http.StatusOK {
				// Indicate that we received an OK from the vault.
				m.Lock()
				resp[url] = true
				m.Unlock()
			} else {
				errText := fmt.Sprintf("%v", err)
				assert.AlwaysOrUnreachable("HTTP Status is never OK when receiving a Post error",
					(err != nil) || (r.StatusCode != http.StatusOK),
					Details{"err": errText, "httpStatus": r.StatusCode})

				// This could include a failure to connect or a timeout during the update.
				glog.Warningf("Error setting vault %s value to %s: %v", vault, string(body), err)
			}
		}(&m, vault, body, resp)
	}
	// Wait for all the connections to complete/timeout/fail.
	wg.Wait()
}

// Check if this number represents a majority of the vaults, where majority has to be >50%.
func (s *ControlServer) hasMajority(count int) bool {
	assert.Always("Control service: determine if there is a majority", true, nil)
	assert.Always("Control service: majority is always expected to be positive", count > 0, Details{"count": count})
	assert.Always("Control service: there are vaults known to the service", len(s.Vaults) > 0, nil)
	numVaults := len(s.Vaults)
	// By default this division will do the equivalent of math.Floor()
	numForMajority := (numVaults / 2) + 1
	haveEnoughVaults := (count >= numForMajority)
	// We expect both conditions below to be sometimes true
	assert.Sometimes("Control service: there is a majority of vaults", haveEnoughVaults, Details{"count": count, "majorityNeeded": numForMajority})
	assert.Sometimes("Control service: there is not a majority of vaults", !haveEnoughVaults, Details{"count": count, "majorityNeeded": numForMajority})
	// We expect numForMajority to be less than 99
	if numForMajority < 99 {
		assert.Unreachable("Control Service: expected failure as we expect the numForMajority to be less than 99 sometimes", Details{"majorityNeeded": numForMajority})
	}
	return haveEnoughVaults
}

func main() {
	fmt.Print("Control Server booting...\n")
	assert.Always("Control service: service started", true, nil)
	portPtr := flag.Int("port", 8000, "Port on which to listen for requests")
	vaultsPtr := flag.String("vaults", "", "Comma-separated list of vaults")
	flag.Parse()
	s := NewControlServer(*vaultsPtr)
	lifecycle.SetupComplete()
	assert.Always("Control service: setup complete", true, nil)
	err := http.ListenAndServe(fmt.Sprintf(":%d", *portPtr), s.mux)
	if errors.Is(err, http.ErrServerClosed) {
		assert.Unreachable("Control service: closed unexpectedly", Details{"error": err})
		fmt.Printf("server closed\n")
	} else if err != nil {
		assert.Unreachable("Control service: did not start", Details{"error": err})
		fmt.Printf("error starting server: %s\n", err)
		os.Exit(1)
	}
}
